drivers/bluetooth/hci/userchan.c

```diff
@@ -0,0 +1,214 @@
+/* userchan.c - HCI User Channel based Bluetooth driver */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <init.h>
+#include <misc/util.h>
+#include <misc/byteorder.h>
+
+#include <errno.h>
+#include <stddef.h>
+#include <poll.h>
+#include <errno.h>
+#include <sys/socket.h>
+#include <string.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/hci_driver.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#include "common/log.h"
+
+#define BTPROTO_HCI      1
+struct sockaddr_hci {
+	sa_family_t     hci_family;
+	unsigned short  hci_dev;
+	unsigned short  hci_channel;
+};
+#define HCI_CHANNEL_USER 1
+
+#define SOL_HCI          0
+
+#define H4_NONE          0x00
+#define H4_CMD           0x01
+#define H4_ACL           0x02
+#define H4_SCO           0x03
+#define H4_EVT           0x04
+
+static BT_STACK_NOINIT(rx_thread_stack, CONFIG_BT_RX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+
+static int uc_fd = -1;
+
+static struct net_buf *get_rx(const u8_t *buf)
+{
+	if (buf[0] == H4_EVT && (buf[1] == BT_HCI_EVT_CMD_COMPLETE ||
+				 buf[1] == BT_HCI_EVT_CMD_STATUS)) {
+		return bt_buf_get_cmd_complete(K_FOREVER);
+	}
+
+	if (buf[0] == H4_ACL) {
+		return bt_buf_get_rx(BT_BUF_ACL_IN, K_FOREVER);
+	} else {
+		return bt_buf_get_rx(BT_BUF_EVT, K_FOREVER);
+	}
+}
+
+static bool uc_ready(void)
+{
+	struct pollfd pollfd = { .fd = uc_fd, .events = POLLIN };
+
+	return (poll(&pollfd, 1, 0) == 1);
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	BT_DBG("started");
+
+	while (1) {
+		static u8_t frame[512];
+		struct net_buf *buf;
+		ssize_t len;
+
+		if (!uc_ready()) {
+			k_sleep(K_MSEC(20));
+			continue;
+		}
+
+		BT_DBG("calling read()");
+
+		len = read(uc_fd, frame, sizeof(frame));
+		if (len < 0) {
+			BT_ERR("Reading socket failed, errno %d", errno);
+			k_yield();
+			continue;
```

Comment:
is this something you can actually recover from?

Reply:
At least in some cases yes, the syscall might return due to an interrupt and set errno = EINTR.